src/compiler/interface.jl

```diff
@@ -61,12 +64,47 @@ end
 
 Base.adjoint(f::Function) = x -> gradient(f, x)[1]
 
+"""
+    withgradient(f, args...)
+    withgradient(f, ::Params)
+
+Returns both the value `f(args...)` and the [`gradient`](@ref), 
+`∂f/∂x` for each argument `x`, as a named tuple.
+With imiplicit parameters, the value is `f()`.
+
+```jldoctest; setup=:(using Zygote)
+julia> y, ∇ = withgradient(/, 1, 2)
+(val = 0.5, grad = (0.5, -0.25))
+
+julia> ∇ == gradient(/, 1, 2)
+true
+```
+"""
+function withgradient(f, args...)
+  y, back = pullback(f, args...)
+  (val=y, grad=back(sensitivity(y)))
+end
+
 # Param-style wrappers
 
-# TODO store ids only
+"""
+    gradient(() -> loss(), ::Params) -> Grads
+
+Gradient with implicit parameters. Returns a container, from which
+`grads[W]` extracts the gradient with respect to some array `W`,
+if this is among those being tracked, for example via `Params([W, A, B])`.
```

Comment:
```suggestion
Gradient with implicit parameters.
Returns a container indexable by the arrays in `Params` (see examples).

```jldoctest
julia> A, B = ones(3, 2), ones(3, 3);

julia> ps = Params([A, B]);

julia> grads = gradient(() -> sum(A) + sum(B), ps)
Grads(...)

julia> grads[A] == ones(3, 2)
true
```

```

Might need to tweak the output to get the test to pass. I wrote this on mobile.

Reply:
Wasn't sure how big to go here. The other `gradient` docsting does not give examples.

Reply:
See if you like my example after update.

Reply:
Yeah the new docstrings with examples all look good